packages/ckeditor5-engine/src/conversion/downcastdispatcher.js

```diff
@@ -636,6 +804,38 @@ function shouldMarkerChangeBeConverted( modelPosition, marker, mapper ) {
 	return !hasCustomHandling;
 }
 
+function getEventName( type, data ) {
+	const name = data.item.name || '$text';
+
+	return `${ type }:${ name }`;
+}
+
+function getParentElementFromChange( entry ) {
+	const { range, position, type } = entry;
+
+	return type === 'attribute' ? range.start.nodeAfter : position.parent;
+}
+
+function walkerValueToEventData( value ) {
+	const item = value.item;
+	const itemRange = Range._createFromPositionAndShift( value.previousPosition, value.length );
+
+	return {
+		item,
+		range: itemRange
+	};
+}
+
+function elementOrTextProxyToView( item, mapper ) {
+	if ( item.is( 'textProxy' ) ) {
+		const mappedPosition = mapper.toViewPosition( Position._createAt( item.parent, item.startOffset ) );
+
+		return mappedPosition.parent.is( '$text' ) ? mappedPosition.parent : null;
```

Comment:
Are you sure this will always work? I didn't analyze it but it looks suspicious. Does mapper always prefer a position inside a text node? I wonder when it returns null (cause AFAICS you expect it) and if you are ready to handle it.